### PR TITLE
Add patient language preference

### LINKED ISSUE #1196: Add patient preferred language
User Story (from a real PHD!):

"I did want to ask if anything can be done about the language variable. For positives, this finding is really important to be able to assign the eventual case to an appropriate case investigator, and we find having this information at the time of reporting dramatically cuts down on hours of staff time due to being able to avoid an interpreter (on average 6 hours/case!)."

Add dropdown field titled "Primary language" to Profile (end of "General information" section)

Tracking responses from health departments here: https://docs.google.com/document/d/1bfPe6UfYmydrRwiRwk7WHn0ffsmGLTZXHg8yB6AnKOw/edit#

Note:
- Default field to "English"
- First 2 options should be "English", then "Spanish". After that, languages should be listed alphabetically
- Use [USWDS combo box component](https://designsystem.digital.gov/components/combo-box/) (same component used for tribal affiliation ticket [#268](https://app.zenhub.com/workspaces/simplereport-better-data-team-605b43d51ea9f700119a48ee/issues/cdcgov/prime-simplereport/268)) 
- Adding dropdown options below once confirmed

### PR DESCRIPTION
**Pull request by nickscialli-usds**
## Related Issue or Background Info

- Resolves #1196

## Changes Proposed

- Adds migration and field on backend (as string)
- Adds field on frontend as string union
- Not sure if we need stricter validation on the backend for this field

### CODE REVIEW

File: frontend/src/config/constants.ts
Comment (jeskew-gov) [inline frontend/src/config/constants.ts:68]: I don't love that we have to type this full list twice. Could this be a map instead of a list? Not sure what the value would be (maybe just `true`?), but that would let you define `Language` as `keyof languages`

Comment (nullflux) [reply frontend/src/config/constants.ts:68]: The trick here is that we're defining them as _types_ in addition to constants, and we use this same pattern for Ethnicity, Gender, YesNo, and Role – it's just less offensive there because there are fewer items in the enumerations.

I'd prefer just using enums (for all of these things!), but that bucks the existing pattern. It _would_ make my separate work w/ Tribal Affiliations much less gross / repetitive.

What do you think about enums here @jeskew-gov?

Comment (jeskew-gov) [reply frontend/src/config/constants.ts:68]: Enums could work, but do you still need to type every member twice? E.g.,

```typescript
enum Language {
  English = "English",
  Spanish = "Spanish",
  ...
}
```

 It's been a while since I used them in Typescript, but I recall that if you didn't provide a string initializer, the enum would default to serializing them as integers

Comment (nullflux) [reply frontend/src/config/constants.ts:68]: Yes, that's correct. I think another option would be to use a const array with a typeof, so something like:

```
const languages = ["English, "Spanish", ...];
type Language = typeof languages[number]
```

Comment (jeskew-gov) [reply frontend/src/config/constants.ts:68]: Nice! I hadn't seen `const` assertions before, but I think that's exactly what we want.

Comment (nullflux) [reply frontend/src/config/constants.ts:68]: @jeskew-gov I played around with this a bit and talked offline with @TimBest: unfortunately even the const assertion option won't work for what we're trying to accomplish with the constants _and_ the type definitions together.

There's not a good way to DRY this up right now

**Review by nathancrtr — APPROVED** [review]
Looks good 👍

**Comment by benwarfield-usds** [comment]
I'd like to talk about this, actually—we now have testResultDelivery and preferredLanguage as columns on the `person` table, with both being saved perpetually as part of the time-of-test record (which seems a bit odd). These seem like, well... preferences. Not part of the person. Though obviously the preferred language is something we would expect staff to always have access to (maybe not so much the preferred results delivery?).

So, question for folks: do we need to split these things out of the `person` table? If not, how many more non-demographic objects to we add before we do need to?

**Comment by nullflux** [comment]
@benwarfield-usds That's a great question. Pulling these fields off of Person simplifies the maintenance of them, and I'd rather do it sooner than later if it's going to result in some number of TestEvents existing in production with serialized Persons that have extraneous fields on them, persisted as an audit in perpetuity.

This also impacts Tribal Affiliation, but _not_ Multiple Phone Numbers, which are the other bodies of work in-flight that affect Person. I'll follow-up on Slack to talk about the "how" & preferred data model

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/PatientPreferences.java
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/PatientPreferences.java:13]: Potentially worth trying jimmy this around to use `@MapsId @JoinColumn(name="internal_id")`? Not sure what we would have to do to inheritance to make that work, though.

File: backend/src/main/java/gov/cdc/usds/simplereport/db/repository/PatientPreferencesRepository.java
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/repository/PatientPreferencesRepository.java:8]: I suspect you always have the Person object here, so possibly just using `findByPerson` would be better?

File: backend/src/main/resources/db/changelog/db.changelog-master.yaml
Comment (benwarfield-usds) [inline backend/src/main/resources/db/changelog/db.changelog-master.yaml:1685]: Use the aliases for these unless we can't (in which case we should do some upstream clean-up to get our non-scalar alias count down so that new code can be clean).

File: backend/src/main/resources/db/changelog/db.changelog-master.yaml
Comment (benwarfield-usds) [inline backend/src/main/resources/db/changelog/db.changelog-master.yaml]: Alternate view on this would be to use the user who last modified the `person` record.

File: backend/src/main/resources/db/changelog/db.changelog-master.yaml
Comment (benwarfield-usds) [inline backend/src/main/resources/db/changelog/db.changelog-master.yaml:1737]: how about `WHERE test_result_delivery IS NOT NULL`?

**Review by benwarfield-usds — COMMENTED** [review]
This looks sane, but we might want to be a little more cautious with the migration. The goal state we're aiming for is that code can be rolled back and then forward again after a migration is run without needing to roll back the migration, and this doesn't meet that (`Person` would fail to validate). I'm not sure if we want to draw that line in the sand here and now, but I think it's worth thinking about. It's pretty easy to just not run the drop-column migration yet—slightly harder if we want to make sure that updates to that column by not-yet-updated code are respected when we roll forward again.

File: backend/src/main/java/gov/cdc/usds/simplereport/db/repository/PatientPreferencesRepository.java
Comment (nullflux) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/repository/PatientPreferencesRepository.java:8]: In the QueueMutationResolver's addPatientToQueue and updateTimeOfTestQuestions we _only_ have a patientId, and I didn't want to unnecessarily add a getPatient call there. Otherwise, yes, related code could just use `findByPerson`. I'll add the additional signature to the interface and remove the couple of unnecessary `getInternalId` calls, though

File: backend/src/main/resources/db/changelog/db.changelog-master.yaml
Comment (nullflux) [reply backend/src/main/resources/db/changelog/db.changelog-master.yaml]: So: leave `created_at` as the migration user, and set `updated_by` to the existing `updated_by` on Person?

**Comment by nullflux** [comment]
@benwarfield-usds: 
> This looks sane, but we might want to be a little more cautious with the migration ... It's pretty easy to just not run the drop-column migration yet.

This is exactly what I've done for now; it ensures backwards compatibility in the event of a software rollback, and it will be plenty clear enough in a future release to remove the comments on that changeset to let it roll

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/PatientPreferences.java
Comment (nullflux) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/model/PatientPreferences.java]: Done :)

Didn't have to mess with inheritance at all

File: backend/src/main/resources/db/changelog/db.changelog-master.yaml
Comment (nullflux) [reply backend/src/main/resources/db/changelog/db.changelog-master.yaml]: Thank you! Done

File: backend/src/main/java/gov/cdc/usds/simplereport/api/patient/PatientDataResolver.java
Comment (nullflux) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/patient/PatientDataResolver.java:32]: This is true, and definitely would be problematic if it happened. I don't want to refactor this to live on a separate entity in GraphQL and require separate requests from the front end at this time.. is there some action you'd recommend? @benwarfield-usds

File: backend/src/main/resources/db/changelog/db.changelog-master.yaml
Comment (benwarfield-usds) [inline backend/src/main/resources/db/changelog/db.changelog-master.yaml:1703]: I know I asked for this, but I think I was tired and spitballing—it probably makes more sense to use `migrations_user_id` as both created-by and updated-by since we're not sure who created or updated the `test_result_delivery` column. The only reason it would make sense to copy the values from `person` is if we want to make some claim about the modification times/users of the person record and how they map to modifications of this new record, and I don't think that is worth the trouble. (I was probably thinking back to the multi-facility migration, where it made sense to attribute the creation and modification times of the organization to the facility, since we were splitting the table vertically and most of the interesting bits were going to the new table.)

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PersonService.java
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PersonService.java:273]: No updateMyPreferredLanguage?

**Review by benwarfield-usds — APPROVED** [review]
:shipit:

File: backend/src/main/java/gov/cdc/usds/simplereport/api/patient/PatientDataResolver.java
Comment (nullflux) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/patient/PatientDataResolver.java]: Prescription removed 👍

File: backend/src/main/resources/db/changelog/db.changelog-master.yaml
Comment (nullflux) [reply backend/src/main/resources/db/changelog/db.changelog-master.yaml:1709]: That would be neat! But yeah this commented out migration _is_ dumb. I'm removing it, and will follow up with a separate PR to be "on hold" until this has been safely released.

File: backend/src/main/resources/db/changelog/db.changelog-master.yaml
Comment (nullflux) [reply backend/src/main/resources/db/changelog/db.changelog-master.yaml:1703]: Gotcha, no worries! I'll set this back to `migrations_user_id`

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PersonService.java
Comment (nullflux) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PersonService.java:273]: Correct – preferred language updates happen via `PersonService.updateMe`, whereas test result delivery preferences are set in AoE questions. I'm going to add a tech debt ticket to clear this up, because (1) it's obfuscated and confusing, and (2) the back end implementation shouldn't follow the awkward quirks of the front end. 

I think there's just enough surface area, now, to warrant a PatientPreferencesService, and for the front end to interact with these values by querying and mutating the PatientPreferences entity directly in the relevant areas.

**Comment by nullflux** [comment]
Sigh. In smoke-testing this before merging it, I've come to realize that `testResultDeliveryPreference` is propagated into the provider view, in the front end, via `GetFacilityQueue`. Adding the `getTestResultDelivery` field onto that query results in _exactly_ the N+1 issue that we added comments to warn about, so even _before_ #1290 this ticket still requires a little bit of rejiggering before it can be merged

File: backend/src/main/java/gov/cdc/usds/simplereport/api/patient/PatientDataResolver.java
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/api/patient/PatientDataResolver.java:52]: Not to belatedly nitpick, but this seems a tad duplicative. 😁

Comment (nullflux) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/patient/PatientDataResolver.java:52]: Addressed in #1338 ;)

**Pull request merged**
